src/puppet.js

```diff
@@ -89,6 +89,122 @@
 
   }
 
+  /**
+   * @constructor
+     */
+  function Reconnector(reconnectAction) {
+    this._reconnect = reconnectAction;
+    this._reset();
+  }
+
+  Reconnector.prototype._reset = function() {
+    this._intervalMs = 1000;
+    this._timeToCurrentReconnectionMs = 0;
+    this._reconnectionPending = false;
+    clearTimeout(this._reconnection);
+    this._reconnection = null;
+  };
+
+  Reconnector.prototype._step = function() {
+    if(this._timeToCurrentReconnectionMs == 0) {
+      console.log("reconnecting now");
+      this._reconnectionPending = false;
+      this._intervalMs *= 2;
+      this._reconnect();
+    } else {
+      console.log('reconnecting in '+(this._timeToCurrentReconnectionMs / 1000) +'s');
+      this._timeToCurrentReconnectionMs -= 1000;
+      setTimeout(this._step.bind(this), 1000);
+    }
+  };
+
+  /**
+   * Notify Reconnector that connection error occurred and automatic reconnection should be scheduled.
+   */
+  Reconnector.prototype.triggerReconnection = function () {
+    if(this._reconnectionPending) {
+      return;
+    }
+    console.log("reconnecting in  "+(this._intervalMs)/1000+"s...");
+    this._timeToCurrentReconnectionMs = this._intervalMs;
+    this._reconnectionPending = true;
+    this._step();
+  };
+
+  /**
+   * Notify Reconnector that connection has been established and there's no need to do further actions.
+   * Resets state of Reconnector
+   */
+  Reconnector.prototype.notifySuccessfulConnection = function() {
+    this._reset();
+  };
+
+  /**
+   * Guarantees some communication to server and monitors responses for timeouts.
+   * @param sendHeartbeatAction will be called to send a heartbeat
+   * @param onError will be called should a response fail to arrive in `timeoutMs`
+   * @param intervalMs if no request will be sent in that time, a heartbeat will be issued
+   * @param timeoutMs shuold a reponse fail to arrive in this time, `onError` will be called
+   * @constructor
+     */
+  function Heartbeat(sendHeartbeatAction, onError, intervalMs, timeoutMs) {
+    this._send = sendHeartbeatAction;
```

Comment:
I didn't mean IIFE but moving methods definitions into constructor, or making properties public.

Our guidelines says not to use `_` in any other case. But it does not say that we should use just naming convention instead of real protection of methods and variables.